common/src/main/java/org/apache/celeborn/common/client/MasterClient.java

```diff
@@ -226,14 +225,23 @@ private void resetRpcEndpointRef(@Nullable RpcEndpointRef oldRef) {
    *     cannot be obtained.
    * @return non-empty RpcEndpointRef.
    */
-  private RpcEndpointRef getOrSetupRpcEndpointRef(AtomicInteger currentIndex) {
+  private RpcEndpointRef getOrSetupRpcEndpointRef(AtomicInteger currentIndex, int currentAttempt) {
     RpcEndpointRef endpointRef = rpcEndpointRef.get();
+
+    List<String> activeMasterEndpoints = masterEndpointResolver.getActiveMasterEndpoints();
+    // If endpoints are updated by MasterEndpointResolver, we should reset the currentIndex to 0.
+    // This also unset the value of updated, so we don't always reset currentIndex to 0.
+    if (masterEndpointResolver.getUpdatedAndReset()) {
+      currentIndex.set(0);
+      maxRetries = Math.max(maxRetries, currentAttempt + activeMasterEndpoints.size());
```

Comment:
Why does this change `maxRetries`? IMO, the `maxRetries` is global in `MasterClient` and does not need to depend on `currentAttempt`. Any missing?

Reply:
Currently maxRetries is set to `maxRetries = max(masterEndpoints.size(), conf.masterClientMaxRetries())` which means client wants to try all the available master endpoint atleast once.

Let's say if we only have one attempt remaining to connect with master and we get a fresh list of master endpoints from resolver IMO we should try all of those atleast once to keep the behaviour almost same. That why i made this change.

Reply:
IMO, it would be better to retain maxRetries as is. If we change maxRetries to a larger value, subsequent retries will use the increased number of attempts, which will alter the original behavior. For example:
Time 1.maxRetries=3
Time 2.retry when currentAttempt=2, activeMasterEndpoints=3, then maxRetries=5
Time 3.retry when currentAttempt=4, activeMasterEndpoints=3, then maxRetries=7
Time 4.retry when currentAttempt=6, activeMasterEndpoints=3, then maxRetries=9
....
 
If the resolver continues to change, the number of maxRetries might increase significantly, leading to exceedingly high retry attempts in subsequent iterations.

And for every endpoint updates, we should never exceed the maxRetries. I think we can use a temp variable in while loop, and check endpoints whether updated before `getOrSetupRpcEndpointRef`, then we can decide increase the temp variable.

Reply:
Agreed @RexXiong, For now i'll retain the value maxRetries as it is then.

I'll raise a followup PR for enhance the logic for resolver updates.

Reply:
updated the diff.